Fix for issue 936 - make sure container registrations are available in BeforeScenario hooks

alexvv opened the pull request:
Fix XUnit ITestOutputhelper and MSTest TestContext container registrations so that they are available in BeforeScenario hooks.

SabotageAndi commented on TechTalk.SpecFlow/Infrastructure/ITestExecutionEngine.cs:
```diff
@@ -16,6 +14,7 @@ public interface ITestExecutionEngine
         void OnFeatureEnd();
 
         void OnScenarioStart(ScenarioInfo scenarioInfo);
+        void OnScenarioStart<T>(ScenarioInfo scenarioInfo, T instanceToAddToScenarioContainer) where T : class;
```
We don't need the type parameter for registering the instance.
What about changing to only one method like: 

    void OnScenarioStart(ScenarioInfo scnearioInfo, object instanceToAddToScenarioContainer = null);

So also the logic duplication what happens in OnScenarioStart is solved.

alexvv replied:
Hi @SabotageAndi, for now I've added some integration tests, and improved a unit test. For OnScenarioStart<T>(..) and other ideas see comments below (https://github.com/techtalk/SpecFlow/pull/1093#issuecomment-377219028).

alexvv replied:
(Copy pasted from overall comments):
Unfortunately, we _do_ need the type parameter in OnScenarioStart. I did a quick test with only using an 'object', but then we can only register its concrete type in the container. Then when resolving an interface (ITestOuthelper) implemented by that object, the BoDi container will give an exception. We could however do something like this:
`void OnScenarioStart(ScenarioInfo scenarioInfo, KeyValuePair<object, Type>? instanceToAdd);`

Or maybe even use a fancy named value tuple, but then test assemblies will become dependent on the System.ValueTuple package. Also I'm not sure if there is a System.CodeDom thing that can generate named value tuples. And KeyValuePair has the advantage that it can be easily changed to an IDictionary, if for some reason we want to register more test-framework specific instances in the future.

SabotageAndi replied:
Ok, I wasn't aware of this behaviour of BoDi. So I am fine with this overload. It's also a little better for compability (see http://thebillwagner.com/Blog/Item/2010-08-23-DegreesofBreakingChangesOptionalParametersinC40)

SabotageAndi commented:
Because of the needed change on the TestExecutionEngine, I am not yet sure if we could release this in the 2.3.* version or need to wait for 2.4/3.0.
Depends on the implementation of OnScenarioStart on the TestExecutionEngine.

alexvv commented:
Ok, I see there are some other related integration tests in TechTalk.SpecFlow.Specs, so I'll add some integration tests there.

Unfortunately, we _do_ need the type parameter in OnScenarioStart. I did a quick test with only using an 'object', but then we can only register its concrete type in the container. Then when resolving an interface (ITestOuthelper) implemented by that object, the BoDi container will give an exception. We could however do something like this:
`void OnScenarioStart(ScenarioInfo scenarioInfo, KeyValuePair<object, Type>? instanceToAdd);`

Or maybe even use a fancy named value tuple, but then test assemblies will become dependent on the System.ValueTuple package. But I'm not sure if there is a System.CodeDom thing that can generate named value tuples. KeyValuePair also has the advantage that it can be easily changed to an IDictionary, if for some reason we want to register more test-framework specific instances in the future.

I also have an idea for reducing duplication in the generators code. If we add something like this to IUnitTestGeneratorProvider:
`KeyValuePair<object, Type>? GetInstanceToAddToScenarioContainer(); // returns null when no registrations are necessary`
the we can use that in UniteTestFeatureGenerator.SetupScenarioInitializeMethod to generate the appropriate scenario container registration (and it is not needed anymore in the MsTest and XUnit generators).

What are your thoughts about this?

alexvv commented:
Hi @SabotageAndi, have you seen my comments on your review? Just checking you're not drowned in github notifications ;) Anyway, for now I've added some integration tests in my PR - we'll still need to discuss OnScenarioStart.

SabotageAndi commented:
@alexvv I saw it, but didn't had time to reply. Thanks for the integration tests, I will have a look at them tomorrow.

alexvv commented:
@SabotageAndi Ok thanks, take your time and let me know if I need to change something. Btw I had a merge from upstream master in my commits, which I fixed recently. I also edited my earlier general remarks comment.

SabotageAndi commented on Tests/TechTalk.SpecFlow.Specs/Features/MsTestProvider.feature:
```diff
@@ -2,181 +2,210 @@
+        [Binding]
+        public class StepsWithTestContextAndBeforeScenario
+        {
+            private Microsoft.VisualStudio.TestTools.UnitTesting.TestContext _testContext;
```
the namespace shouldn't be needed here.

alexvv replied:
Ok, will be fixed

SabotageAndi commented on Tests/TechTalk.SpecFlow.Specs/Features/XUnit2Provider.feature:
```diff
@@ -82,150 +82,177 @@ Scenario: Should be able to ignore all Scenarios in a feature
+        | Succeeded |
+        | 1         |
 
+Scenario: Should be able to log custom messages using context injection
```
Is this scenario still needed? "Should be able to access ITestOutputHelper using injection and BeforeScenario hook" tests it much better.

We don't test for the output in this scenario. It's a test if _output is null, but with an ugly error message. ;-)

alexvv replied:
Well, you could argue that a different type is used for injection - which is ScenarioContext instead of ITestOutputHelper. Also a 'When' step is executed, which I left out in the other scenario. 
But I can combine them to 1 scenario by using a constructor with 2 arguments (ScenarioContext and ITestOutputhelper), and assert them.

SabotageAndi commented:
@alexvv & @gasparnagy 
This PR is for XUnit and MSTest. What about NUnit? The [TestContext](https://github.com/nunit/docs/wiki/TestContext) has a lot of static properties/methods, but they also have a TestContext.CurrentContext property. We could inject this also to the ScenarioContext so that user would be able to get it via DI and don't have to use the static properties/methods.

What are you thinking?

alexvv commented:
@SabotageAndi and @gasparnagy  I think the NUnit TestContext injection is a good idea, makes it somewhat consistent with the other (most popular) test frameworks. I'll deliver a proposal in my next commit.

alexvv commented:
Hi @SabotageAndi and @gasparnagy, could you take a look at my latest [commit](https://github.com/techtalk/SpecFlow/pull/1093/commits/246bdcae0ea8d3cba7e70bf82e5038f58c62857c) ? I've processed your remarks and I've added support for NUnit TestContext.CurrentContext injection to this PR. 
Furthermore, I've also removed the unit tests `XUnit2TestGeneratorProviderTests.Should_register_testOutputHelper_on_scenario_setup`  and   `MsTestGeneratorProviderTests.MsTestGeneratorShouldRegisterTestContextOnScenarioSetup`, because they're more kind of integration-tests and the functionality is also covered by the (new) full integration tests in `MsTestProvider.feature` and `XUnit2Provider.feature`. Let me know if you disagree :)

SabotageAndi commented:
@alexvv looks good for me.
@gasparnagy any thoughts on your side?

gasparnagy commented:
@alexvv it is good that you looked into this problem and generally I think this could be a good direction to solve the things. Based on what I have seen in the PR, I was thinking if we could do this somewhat in a simpler way. Let me summarize what my idea is. //cc @SabotageAndi 

So the core problem is the following:

1. The TestExecutionEngine.OnScenarioStart does two things: a) it initializes the scenario context b) it fires the before scenario events
2. Some (more and more) test runners need to inject additional object(s) to the scenario context before the scenario events
3. The additional objects are only available from the generated test class, so to be able to reach the scenario context, they have to be passed through the test runner and the test execution engine

Based on that there are two options I could imagine:

1. Split the TestExecutionEngine.OnScenarioStart (and the related method on the test runner) into two methods: OnScenarioInitialize and OnScenarioStart. The first would setup the scenario context the other would fire the events, and we would generate two method invokes to the generated code. In this case, the unit test providers could somehow inject code in between these codes (through a new method on the unit test generator interface that does not do anything by default).
2. Somehow let the information flow down to the TestExecutionEngine, similarly like you did. But in this case, I would try to avoid adding these test output helper objects to the interfaces of SpecFlow, as SpecFlow does not know about them (and what if we need two of them?). Instead, the parameter you add to the OnScenarioStart could be a delegate (`Action<ScenarioContext>` named e.g. `customzeScenarioContext`), so that any test runner generator can decide what to do there and we could also avoid the generic type parameter, so this could be simply an optional parameter, defaults to null (means, no customization is necessary). I would also try to avoid introducing the traits (these soould be used more for indicating capabilities not to alter the generation flow) and simply add a method to the generator interface that returnes the code dom expression that should be used as a second parameter (by default returning a null-expression).


I think this second idea is not very far to what you have created, would be interesting to see if it would make this entire stuff clearer. 

Also you should think about the first option, if you like that more. For me both would be fine.

(This is a tricky issue and since the changes are affecting the core interfaces of SpecFlow, let's try to think it over carefully.)

alexvv commented:
@gasparnagy Thanks for your comments. I think the first option with splitting up OnScenarioStart is the nicest option, this way no (optional) action callbacks are needed and we can easily get rid of the new traits as well (using empty methods in the generators if nothing is needed for OnScenarioInitialize). I'll look into this, but it will take some time.

alexvv commented:
Hi @gasparnagy, it's been a while - but I've committed a new implementation where TestExecutionEngine.OnScenarioStart has been split up. Could you take a look at it please?

gasparnagy commented:
@alexvv I think this is pretty clean like that. @SabotageAndi do you see any problem with this approach?

For the code, my only request is to remove the version changes in the `.nuspec` files from the PR.

SabotageAndi commented:
@gasparnagy I am fine with the changes.

SabotageAndi commented:
@alexvv I would like to merge this into for SpecFlow 2.4.
Could you remove the nuspec changes from the PR until Thursday?

alexvv commented:
Hi @SabotageAndi, that's great news. I've removed the nuspec changes from the commit (and it's still a single commit).

SabotageAndi commented:
Merged.
@alexvv They could have been separate commits. I squash all PRs into a single commit.
